### PR TITLE
Doc fixes

### PR DESCRIPTION
**Pull request by mpariente**
Trying to fix the docstrings to render nicely on 0.4.0 release. 

I'm struggling with functions that return multiple things, the docstring is never parsed right when it starts by the return type. 

### Comparing docstrings and visual
1. Looks ok but it doesn't have the dots like in the args.
```python
    def trial(self, pair_wise_losses, perm_reduce=None, **kwargs):
        r"""Trial docstring

        Args:
            pair_wise_losses: there is a dot in the beginning.
            perm_reduce: And it's automatic. 

        Returns:
            :class:`torch.Tensor`: The loss corresponding to the best
            permutation of size $(batch,)$.

            :class:`torch.Tensor`: The indices of the best permutations.
        """
```

image

2. Sphinx is joking with me 
```python 
    def trial(self, pair_wise_losses, perm_reduce=None, **kwargs):
        r"""Trial docstring

        Args:
            pair_wise_losses: there is a dot in the beginning.
            perm_reduce: And it's automatic.

        Returns:
            - :class:`torch.Tensor`: The loss corresponding to the best
              permutation of size $(batch,)$.

            - :class:`torch.Tensor`: The indices of the best permutations.
        """
```

image

3. Looks ok but the line is broken
```python
    def trial(self, pair_wise_losses, perm_reduce=None, **kwargs):
        r"""Trial docstring

        Args:
            pair_wise_losses: there is a dot in the beginning.
            perm_reduce: And it's automatic.

        Returns:
            - :class:`torch.Tensor`: The loss corresponding to the best
            permutation of size $(batch,)$.

            - :class:`torch.Tensor`: The indices of the best permutations.
        """
```

image

4. This is looks great, but cannot start with the type. 
```python
    def trial(self, pair_wise_losses, perm_reduce=None, **kwargs):
        r"""Trial docstring

        Args:
            pair_wise_losses: there is a dot in the beginning.
            perm_reduce: And it's automatic.

        Returns:
            - The loss corresponding to the best permutation of size $(batch,)$.
              (:class:`torch.Tensor`: )

            - The indices of the best permutations (:class:`torch.Tensor`:).
        """
```

image

### CODE REVIEW

**Comment by jonashaag** [comment]
Haha yes Sphinx/reST syntax really is horrible.

I think the bullets just aren't supported in Google-style docstrings. This works (NumPy-style):

```
def trial(self, pair_wise_losses, perm_reduce=None, **kwargs):
    r"""Trial docstring

    Returns
    -------
    :class:`torch.Tensor`
        The loss corresponding to the best permutation of size $(batch,)$.

    :class:`torch.Tensor`
        The indices of the best permutations.
    """

```


Btw I found this which I think would be nice for tensors: https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#confval-napoleon_type_aliases

File: asteroid/data/avspeech_dataset.py [hunk 1]
```diff
@@ -120,8 +120,8 @@ class AVSpeechDataset(data.Dataset):
         n_src (int): number of sources.
 
     References
-        - [1]: "Looking to Listen at the Cocktail Party: A Speaker-Independent Audio-Visual
-          Model for Speech Separation" Ephrat et. al https://arxiv.org/abs/1804.03619
+        "Looking to Listen at the Cocktail Party: A Speaker-Independent Audio-Visual
+        Model for Speech Separation" Ephrat et. al https://arxiv.org/abs/1804.03619
```
Comment (jonashaag) [inline asteroid/data/avspeech_dataset.py:124]:
```suggestion
        [1] "Looking to Listen at the Cocktail Party: A Speaker-Independent Audio-Visual
        Model for Speech Separation" Ephrat et. al https://arxiv.org/abs/1804.03619
```

(?)

File: asteroid/complex_nn.py [hunk 2]
```diff
@@ -176,8 +176,8 @@ def bound_complex_mask(mask: ComplexTensor, bound_type="tanh"):
             "tanh"/"bdt" (default), "sigmoid"/"bdss" or None/"bdt".
 
     References
```
Comment (jonashaag) [inline asteroid/complex_nn.py:178]: Do you know why we have `:` at the end of `Args:` but not for `References`?

File: asteroid/dsp/overlap_add.py [hunk 3]
```diff
@@ -248,22 +246,22 @@ def unfold(self, x):
...
-            x: (:class:`torch.Tensor`): spliced feature tensor of shape
-                (batch, channels, chunk_size, n_chunks).
-            output_size: (int, optional): sequence length of original feature tensor.
-                If None, the original length cached by the previous call of `unfold`
-                will be used.
+            x (:class:`torch.Tensor`): spliced feature tensor of shape
+                $(batch, channels, chunksize, nchunks)$.
+            output_size (int, optional): sequence length of original feature tensor.
+                If None, the original length cached by the previous call of
+                :func:`~overlap_add.DualPathProcessing.unfold` will be used.
```
Comment (jonashaag) [inline asteroid/dsp/overlap_add.py:259]: I believe this could also be written as

```
:meth:`unfold`
```

File: asteroid/losses/pit_wrapper.py [hunk 4]
```diff
@@ -229,13 +228,12 @@ def reorder_source(source, batch_indices):
         """Reorder sources according to the best permutation.
 
         Args:
-            source (torch.Tensor): Tensor of shape [batch, n_src, time]
-            batch_indices (torch.Tensor): Tensor of shape [batch, n_src].
+            source (torch.Tensor): Tensor of shape :math:
```
Comment (jonashaag) [inline asteroid/losses/pit_wrapper.py:231]: copy paste error?

File: asteroid/losses/pit_wrapper.py [hunk 5]
```diff
@@ -244,24 +242,23 @@ def reorder_source(source, batch_indices):
...
     @staticmethod
     def find_best_perm_factorial(pair_wise_losses, perm_reduce=None, **kwargs):
-        """Find the best permutation given the pair-wise losses by looping
+        r"""Find the best permutation given the pair-wise losses by looping
         through all the permutations.
 
         Args:
             pair_wise_losses (:class:`torch.Tensor`):
-                Tensor of shape [batch, n_src, n_src]. Pairwise losses.
+                Tensor of shape :math:. Pairwise losses.
```
Comment (jonashaag) [inline asteroid/losses/pit_wrapper.py:250]: copy paste error?

File: asteroid/masknn/convolutional.py [hunk 6]
```diff
@@ -400,12 +399,12 @@ class DCUNetComplexEncoderBlock(nn.Module):
         stride (Tuple[int, int]): Convolution stride.
         padding (Tuple[int, int]): Convolution padding.
         norm_type (str, optional): Type of normalization to use.
-            See ``asteroid.masknn.norms`` for valid values.
+            See :func:`~asteroid.masknn.norms` for valid values.
```
Comment (jonashaag) [inline asteroid/masknn/convolutional.py:402]:
```suggestion
            See :mod:`~asteroid.masknn.norms` for valid values.
```

It's a module right?

File: asteroid/masknn/convolutional.py [hunk 7]
```diff
@@ -400,12 +399,12 @@ class DCUNetComplexEncoderBlock(nn.Module):
...
         activation (str, optional): Type of activation to use.
-            See ``asteroid.masknn.activations`` for valid values.
+            See :func:`~asteroid.masknn.activations` for valid values.
```
Comment (jonashaag) [inline asteroid/masknn/convolutional.py:404]:
```suggestion
            See :mod:`~asteroid.masknn.norms` for valid values.
```

File: asteroid/masknn/convolutional.py [hunk 8]
```diff
@@ -444,12 +443,12 @@ class DCUNetComplexDecoderBlock(nn.Module):
         stride (Tuple[int, int]): Convolution stride.
         padding (Tuple[int, int]): Convolution padding.
         norm_type (str, optional): Type of normalization to use.
-            See ``asteroid.masknn.norms`` for valid values.
+            See :func:`~asteroid.masknn.norms` for valid values.
```
Comment (jonashaag) [inline asteroid/masknn/convolutional.py:446]:
```suggestion
            See :mod:`~asteroid.masknn.norms` for valid values.
```

File: asteroid/masknn/convolutional.py [hunk 9]
```diff
@@ -444,12 +443,12 @@ class DCUNetComplexDecoderBlock(nn.Module):
...
         activation (str, optional): Type of activation to use.
-            See ``asteroid.masknn.activations`` for valid values.
+            See :func:`~asteroid.masknn.activations` for valid values.
```
Comment (jonashaag) [inline asteroid/masknn/convolutional.py:448]:
```suggestion
            See :mod:`~asteroid.masknn.norms` for valid values.
```

File: docs/README.md [hunk 10]
```diff
@@ -25,3 +25,11 @@ from the `docs/` folder, that'll be easier.
 ```bash
 alias run_docs='make clean; make html; firefox build/html/index.html'
 ```
+
+
+### Writing good docstrings
+
+- Start with [RST and Sphinx CheatSheet](https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html)
+- Linking to method in a class `:func:~mymodule.myclass.myfunc`.
```
Comment (jonashaag) [inline docs/README.md:33]: That would be `:meth` actually, see here https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects (you can omit the `:py:`)

Comment (jonashaag) [reply docs/README.md:33]: But I think you can use any tag (role) and it works anyways

**Comment by jonashaag** [comment]
Btw what's that `$...$` syntax? Never seen it (only in LaTeX and Markdown)

File: asteroid/complex_nn.py (hunk 2 above)
Comment (mpariente) [reply asteroid/complex_nn.py:178]: `Args:` is detected by Sphinx and it handles it well. 
And if I add `:` to `Reference`, it looks like this: 
image

I really don't understand the tool, so when I find something that works, I superstitiously stick to it.

File: asteroid/dsp/overlap_add.py (hunk 3 above)
Comment (mpariente) [reply asteroid/dsp/overlap_add.py:259]: Yes, it works. Thanks a lot.

File: asteroid/losses/pit_wrapper.py (hunk 5 above)
Comment (mpariente) [reply asteroid/losses/pit_wrapper.py:250]: `sed` error ^^
Thanks

File: asteroid/masknn/convolutional.py (hunk 6 above)
Comment (mpariente) [reply asteroid/masknn/convolutional.py]: Sorry, I committed that. And thanks for the suggestion.

**Comment by mpariente** [comment]
Thanks a lot for your feedback !

> Btw what's that `$...$` syntax? Never seen it (only in LaTeX and Markdown)

No idea, but it works :sparkles: :joy:

**Comment by mpariente** [comment]
What's your call on the multiple output function then? 
I'm ok not to start with the type, but this seems like black magic.

**Comment by jonashaag** [comment]
I really don't know. Here's what the internet says... 

https://stackoverflow.com/questions/29221551/can-sphinx-napoleon-document-function-returning-multiple-arguments
http://web.archive.org/web/20200702074210/https://bitbucket.org/birkenfeld/sphinx-contrib/issues/111/napoleon-multiple-returns-indentations

So looks like Google style just doesn't support this.

I guess the best way to implement it would be using lists (like in your example 2) but not sure how much work it is to get this into Sphinx.

Also found this, not sure if related https://github.com/sphinx-contrib/napoleon/issues/4

**Comment by mpariente** [comment]
Thanks for the pointers ! 

Finally found a way that worked: 
image

```python
    def trial(self, pair_wise_losses, perm_reduce=None, **kwargs):
        r"""Trial docstring

        Args:
            pair_wise_losses: there is a dot in the beginning.
            perm_reduce: And it's automatic.

        Returns
            - :class:`torch.Tensor`:
              The loss corresponding to the best permutation of size $(batch,)$. and
              if I keep typing? It works? 

            - :class:`torch.Tensor`:
              The indices of the best permutations.
        """
```

**Comment by mpariente** [comment]
I tried ` napoleon_type_aliases` and couldn't make it work out of the box. 

By the way, do you know how to automatically have anchor points for each class/function that is documented?

**Comment by jonashaag** [comment]
> By the way, do you know how to automatically have anchor points for each class/function that is documented?

No sorry :( My best advice is to look at other documentations and try to copy their config :-D 

> Finally found a way that worked:

This is nice, if this works out ok I think it would be great to add this to Stack Overflow

**Pull request merged**